web3.js/src/connection.ts

```diff
@@ -2761,80 +2762,257 @@ export class Connection {
     return res.result;
   }
 
+  // async confirmTransactionBlockHeight(
+  //   signature: TransactionSignature,
+  //   latestBlockhash: TransactionBlockhash,
+  //   commitment?: Commitment,
+  // ): Promise<RpcResponseAndContext<SignatureResult>> {
+  //   let decodedSignature;
+  //   try {
+  //     decodedSignature = bs58.decode(signature);
+  //   } catch (err) {
+  //     throw new Error('signature must be base58 encoded: ' + signature);
+  //   }
+
+  //   assert(decodedSignature.length === 64, 'signature has invalid length');
+
+  //   const subscriptionCommitment = commitment || this.commitment;
+  //   let subscriptionId;
+  //   let done = false;
+  //   let response: RpcResponseAndContext<SignatureResult> | null = null;
+
+  //   const confirmTx = new Promise((resolve, reject) => {
+  //     try {
+  //       subscriptionId = this.onSignature(
+  //         signature,
+  //         (result: SignatureResult, context: Context) => {
+  //           subscriptionId = undefined;
+  //           response = {
+  //             context,
+  //             value: result,
+  //           };
+  //           done = true;
+  //           resolve(TransactionStatus.CONFIRMED);
+  //         },
+  //         subscriptionCommitment,
+  //       );
+  //     } catch (err) {
+  //       reject(err);
+  //     }
+  //   });
+
+  //   const checkBlockHeight = async () => {
+  //     const blockHeight = await this.getBlockHeight(commitment);
+  //     return blockHeight;
+  //   };
+
+  //   const expireTx = new Promise((resolve, reject) => {
+  //     (async () => {
+  //       try {
+  //         let currentBlockHeight = await checkBlockHeight();
+  //         while (currentBlockHeight <= latestBlockhash.lastValidBlockHeight!) {
+  //           if (done) return;
+  //           await sleep(1000);
+  //           currentBlockHeight = await checkBlockHeight();
+  //         }
+  //         resolve(TransactionStatus.EXPIRED);
+  //       } catch (error) {
+  //         reject(error);
+  //       }
+  //     })();
+  //   });
+
+  //   try {
+  //     const outcome = await Promise.race([confirmTx, expireTx]);
+  //     switch (outcome) {
+  //       case 'CONFIRMED':
+  //         console.log('Transaction has been confirmed');
+  //         break;
+  //       case 'EXPIRED':
+  //         console.log('Transaction has expired');
+  //         break;
+  //     }
+  //   } finally {
+  //     if (subscriptionId) {
+  //       this.removeSignatureListener(subscriptionId);
+  //     }
+  //   }
+
+  //   if (response === null) {
+  //     throw new Error(`Transaction ${signature} expired`);
+  //   }
+
+  //   return response;
+  // }
+
+  confirmTransaction(
+    signature: {
+      signature: TransactionSignature;
+      latestBlockhash: TransactionBlockhash;
+    },
```

Comment:
I agree. I'm not too sure about how to do this. I think I understand the basic idea of `nonce`s but I haven't actually used one before. This was the next thing I wanted to figure out.

Reply:
I'd say just make sure that you're not painting yourself into a corner. For instance, there's a potential footgun here with the type system once you add an `opts` type for nonce-based confirmations: ([playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAcg9gOwMYQMKIGYEsDmUC8UA3gFBRQKIoBcUAzsAE5YI4A0JAvgNwkmiQoAZVwIAhsACujNJlwFiZeqInSItBs1YdyAGwkQGAIV1wkAawAWYupdoJJAWwBGERhx58Mk5MCyIoJDlGRwAVRjEEOjEkP0QACgiAd3QEbBxaeGRZNPkAH2EVKRlU9IBKRXIsDCh4gHlnACsIWIA6MEY4YC7wCFbrOjqkhAAFTshGUFakMV1dRLEUuXYoAHJKbNWyitJyciCo4EDlhWTS+RtYKhz03j2oAHoHqDGIMDEZCmuoYEsWHDorSUnCgEF0dGguz2BwYx1yeEIZxOlxEOHExRuuDueyeLxk70+ADdZlgACZQZymCyWCC4SxHX7-QHArh8A7YELhSLRWL+BDxIgbGhrDBwOCrThlbiPZ4AcTFpJI7KwnIiURicX5RDoRTUtFWovFbCg+mAhmAJjMVhsdigAEYAEwAZkl0tx8rgiqVwTCap5moFOrRqhk+sNq2NQvUIrFEqlMuEljgkl0pIQqyOrigs1MSQgpIAhEA))

You'll need a type strategy to make sure that people can't supply incorrect config.

One way of doing so is to make the config a tagged union, but then you'll likely need helpers to create these tagged objects, so that people don't have to muck around and do it themselves. Adds a layer of complexity that we maybe don't want. ([playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAcg9gOwMYQMKIGYEsDmUC8UA3gFBRQD6FokAXFAOQKIoC0ARgIYDOEAJqySZcDADRkozZBHrdgAJywIc4gL4BuEiRrQAyrgSdgAV3lpheQqXJUd9BuwA2cJAGsAFhFzvgHHv0ELMQluAyNTGSg5RWVxckcjCDkAIWc3dx53egRjAFt2CHk1TRIMY2RgLEQoIQRseVyAFXlOBG5OJErEAAoWgHd0Otx6eGlB7DwAHyh9HEMTM3HcAEpiCSwMKF7OAYsAOltwaHwTxik2Ll4BWomGVetyGsQ5J6HLKH6lnE1HqAB6P5QAAKZjAnDMkhY0GA7iUOG4ewkqigEEcvDWv1qLxuuAIHx2Xx+jwBwNB4OgADdOI4sHwoE4XB4vDgfFAYXCEUiSKotDiGs1Wu1OlUEN0kGZEqMUF9ugwMHA4Hdlup-oCAOIKvgkPlNFptDpdUXiiCJVKMzzeYDJfx8GVyhViKAARgATABmZbK1VQDVwLXa4T8vVCw3dIihObhMz2eWK0SQ6Qxh2qL0k3TuODGRx8BAMYD06DU5x9fgAQi0ZQqIpqEuAECl5je3XOkWicOWIyhXwxHwgCwQPZs1CO9hbfiugTewV+LbiUA03Ir5WF1WNprSTMt1quMoj8wisgUcPjCTrKQ3GW4WUkeQK8g7s33iwsPbM-cHlGHdEYDPSzJ844BDiODTo8p5JFaF6ZHO5B7lGEBzguqhAA))